runtime/commonMain/src/kotlinx/serialization/json/internal/TreeJsonOutput.kt

```diff
@@ -68,8 +56,13 @@ private sealed class AbstractJsonTreeOutput(
     }
 
     override fun <T> encodeSerializableValue(serializer: SerializationStrategy<T>, value: T) {
-        encodePolymorphically(serializer, value) {
-            writePolymorphic = true
+        if (serializer.descriptor.kind !is PrimitiveKind && serializer.descriptor.kind !== UnionKind.ENUM_KIND || currentTagOrNull != null) {
+            encodePolymorphically(serializer, value) { writePolymorphic = true }
+        } else {
+            JsonTreePrimitiveOutput(json, nodeConsumer).apply {
```

Comment:
IMO, `apply` is not applicable here semantically, because it creates a false sense that you are somehow changing it in `block`, but this is not true. And every indentation level makes code less readable.
E.g.
```
if (serializer.descriptor.kind !is PrimitiveKind && serializer.descriptor.kind !== UnionKind.ENUM_KIND || currentTagOrNull != null) {
    return encodePolymorphically(serializer, value) { writePolymorphic = true }
}
val output = JsonTreePrimitiveOutput(json, nodeConsumer)
output.encodeSerializableValue(serializer, value)
output.endEncode(serializer.descriptor)
```

Reply:
Well, actually, calls to `encode` change output state somehow, so I don't see problems with `apply`. Maybe something like this?

```kotlin
if (/**...**/) {
    encodePolymorphically(serializer, value) { writePolymorphic = true }
} else JsonTreePrimitiveOutput(json, nodeConsumer).apply {
    encodeSerializableValue(serializer, value)
    endEncode(serializer.descriptor)
}
```